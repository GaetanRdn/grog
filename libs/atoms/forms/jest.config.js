module.exports = {
  displayName: 'atoms-forms',
  preset: '../../../jest.preset.js',
  setupFilesAfterEnv: ['<rootDir>/src/test-setup.ts'],
  globals: {
    'ts-jest': {
      tsconfig: '<rootDir>/tsconfig.spec.json',
      stringifyContentPathRegex: '\\.(html|svg)$',
    },
  },
  collectCoverage: true,
  coverageDirectory: '../../../coverage/libs/atoms/forms',
  transform: {
    '^.+\\.(ts|js|html)$': 'jest-preset-angular',
  },
  snapshotSerializers: [
    'jest-preset-angular/build/serializers/no-ng-attributes',
    'jest-preset-angular/build/serializers/ng-snapshot',
    'jest-preset-angular/build/serializers/html-comment',
  ],
};
