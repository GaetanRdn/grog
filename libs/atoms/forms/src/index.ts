export * from './lib/input/input.directive';
export * from './lib/autocomplete/autocomplete.component';
export * from './lib/label/label.component';
export * from './lib/checkbox/checkbox.component';
export * from './lib/select/select.component';
