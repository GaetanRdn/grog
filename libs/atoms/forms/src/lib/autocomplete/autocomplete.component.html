<input
  adrInput
  [value]="displayedValue"
  (focus)="openPanel()"
  (input)="filterOptions($event.target)"
  [disabled]="disabled"
  (blur)="createOption($event.target)"
/>
<div *ngIf="isOpen" class="adr-autocomplete-options">
  <span
    class="adr-option"
    [class.adr-selected]="isSelected(option)"
    *ngFor="let option of displayedValues$ | async; trackBy: trackByFn"
    (click)="select(option)"
    >{{ displayOptionFn(option) }}</span
  >
</div>
