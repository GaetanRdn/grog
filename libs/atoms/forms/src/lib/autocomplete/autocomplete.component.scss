@use 'utils';

@mixin adr-autocomplete($config) {
  adr-autocomplete {
    flex: 1;

    &.adr-opened {
      .adr-input {
        border-bottom-right-radius: unset;
        border-bottom-left-radius: unset;
      }
    }

    .adr-autocomplete-options {
      @include utils.set-typography($config, label);
      @include utils.set-color-values($config, primary);
      position: absolute;
      display: flex;
      flex-flow: column;
      align-items: stretch;
      width: 100%;
      cursor: pointer;
      border-left-width: 2px;
      border-right-width: 2px;
      border-bottom-width: 2px;
      border-style: inset;
      border-bottom-right-radius: 0.25rem;
      border-bottom-left-radius: 0.25rem;

      &:last-child {
        border-bottom-right-radius: 0.25rem;
        border-bottom-left-radius: 0.25rem;
      }

      .adr-option {
        padding: 0.5rem 1rem;

        &:hover {
          @include utils.set-color-values($config, primary-dark);
        }

        &.adr-selected {
          @include utils.set-color-values($config, primary-light);
        }
      }
    }
  }
}
