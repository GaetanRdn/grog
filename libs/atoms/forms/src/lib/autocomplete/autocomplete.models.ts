export type CreateOptionFn<T> = (input: string) => T;

export type DisplayFn<T> = (option: T) => string;

export type IdentityFn<T> = (option: T) => any;

export type OpenOn = 'focus' | 'input';
