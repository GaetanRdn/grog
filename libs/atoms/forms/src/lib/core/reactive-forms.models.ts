export type OnChangeFn<ValueType> = (value: ValueType | null) => void;
export type OnTouchedFn = () => void;
