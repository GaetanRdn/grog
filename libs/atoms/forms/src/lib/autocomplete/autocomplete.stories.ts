import { FormControl, ReactiveFormsModule } from '@angular/forms';
import { action } from '@storybook/addon-actions';
import { Meta, Story } from '@storybook/angular';
import {
  AutocompleteComponent,
  AutocompleteModule,
} from './autocomplete.component';

export default {
  title: 'atoms/forms/autocomplete',
  component: AutocompleteComponent,
  parameters: {
    jest: ['autocomplete.component.spec.ts'],
  },
} as Meta<AutocompleteComponent<string>>;

const template: Story<AutocompleteComponent<string>> = (
  args: AutocompleteComponent<string>
) => ({
  moduleMetadata: {
    imports: [AutocompleteModule],
  },
  props: {
    ...args,
    valueChange: action('valueChange'),
  },
  template: `<gro-autocomplete [value]='value' [openOn]='openOn' [options]='options' [required]='required' [disabled]='disabled' (valueChange)='valueChange($event)'></gro-autocomplete>`,
});

export const basic = template.bind({});
basic.args = {
  options: ['Gaetan', 'Soren', 'Bernard'],
  value: '',
};

interface Person {
  id?: number;
  firstName: string;
  name: string;
}

const objectsValuesTemplate: Story<AutocompleteComponent<Person>> = (
  args: AutocompleteComponent<Person>
) => ({
  moduleMetadata: {
    imports: [AutocompleteModule],
  },
  props: {
    ...args,
    valueChange: action('valueChange'),
    displayOptionFn: (option: Person): string =>
      `${option.firstName} - ${option.name}`,
  },
  template: `<gro-autocomplete [value]='value' [openOn]='openOn' [options]='options' [required]='required' [disabled]='disabled' (valueChange)='valueChange($event)' [displayOptionFn]='displayOptionFn'></gro-autocomplete>`,
});

export const complexValues = objectsValuesTemplate.bind({});
complexValues.args = {
  options: [
    { id: 1, firstName: 'Gaetan', name: 'Redin' },
    { id: 2, firstName: 'Soren', name: 'Redin' },
    { id: 3, firstName: 'Lord', name: 'Voldemor' },
  ],
  value: { id: 2, firstName: 'Soren', name: 'Redin' },
};

const reactiveFormTemplate: Story<AutocompleteComponent<Person>> = (
  args: AutocompleteComponent<Person>
) => ({
  moduleMetadata: {
    imports: [AutocompleteModule, ReactiveFormsModule],
  },
  props: {
    ...args,
    valueChange: action('valueChange'),
    value: null,
    displayOptionFn: (option: Person): string =>
      `${option.firstName} - ${option.name}`,
    control: new FormControl({ id: 2, firstName: 'Soren', name: 'Redin' }),
    identityFn: (p: Person): string | number => p.id ?? p.firstName + p.name,
  },
  template: `<gro-autocomplete [formControl]='control' [openOn]='openOn' [options]='options' [required]='required' (valueChange)='valueChange($event)' [displayOptionFn]='displayOptionFn' [identityFn]='identityFn'></gro-autocomplete>`,
});

export const reactiveForm = reactiveFormTemplate.bind({});
reactiveForm.args = {
  options: [
    { id: 1, firstName: 'Gaetan', name: 'Redin' },
    { id: 2, firstName: 'Soren', name: 'Redin' },
    { id: 3, firstName: 'Lord', name: 'Voldemor' },
  ],
};

const addOptionTemplate: Story<AutocompleteComponent<Person>> = (
  args: AutocompleteComponent<Person>
) => ({
  moduleMetadata: {
    imports: [AutocompleteModule, ReactiveFormsModule],
  },
  props: {
    ...args,
    valueChange: action('valueChange'),
    value: { id: 2, firstName: 'Soren', name: 'Redin' },
    displayOptionFn: (option: Person): string =>
      option.firstName + ' - ' + option.name,
    identityFn: (p: Person): string | number => p.id ?? p.firstName + p.name,
    createOptionFn: (input: string): Partial<Person> => {
      if (input.includes('-')) {
        return {
          firstName: input.split('-')[0].trim(),
          name: input.split('-')[1].trim(),
        } as Person;
      } else {
        return {
          firstName: input.split('-')[0].trim(),
        } as Person;
      }
    },
  },
  template: `<gro-autocomplete [value]='value' [openOn]='openOn' [options]='options' [required]='required' (valueChange)='valueChange($event)' [displayOptionFn]='displayOptionFn' [createOptionFn]='createOptionFn' [identityFn]='identityFn'></gro-autocomplete>`,
});

export const addOption = addOptionTemplate.bind({});
addOption.args = {
  options: [
    { id: 1, firstName: 'Gaetan', name: 'Redin' },
    { id: 2, firstName: 'Soren', name: 'Redin' },
    { id: 3, firstName: 'Lord', name: 'Voldemor' },
  ],
};
