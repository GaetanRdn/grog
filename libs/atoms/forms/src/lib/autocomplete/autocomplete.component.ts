import { CommonModule } from '@angular/common';
import {
  ChangeDetectionStrategy,
  ChangeDetectorRef,
  Component,
  ElementRef,
  EventEmitter,
  forwardRef,
  HostListener,
  Input,
  NgModule,
  Output,
  Provider,
  TrackByFunction,
  ViewChild,
} from '@angular/core';
import { ControlValueAccessor, NG_VALUE_ACCESSOR } from '@angular/forms';
import { AutoUnsubscribe, CoerceBoolean } from '@grorg/decorators';
import { BehaviorSubject, Observable } from 'rxjs';
import { OnChangeFn, OnTouchedFn } from '../core/reactive-forms';
import { InputDirective, InputModule } from '../input/input.directive';
import {
  CreateOptionFn,
  DisplayFn,
  IdentityFn,
  OpenOn,
} from './autocomplete.models';

const AUTOCOMPLETE_VALUE_ACCESSOR: Provider = {
  provide: NG_VALUE_ACCESSOR,
  useExisting: forwardRef(() => AutocompleteComponent),
  multi: true,
};

@Component({
  selector: 'gro-autocomplete',
  // standalone: true,
  host: {
    '[class.gro-opened]': 'isOpen',
    '[class.gro-disabled]': 'disabled',
    '[attr.required]': 'required || null',
    '[attr.disabled]': 'disabled || null',
  },
  templateUrl: './autocomplete.component.html',
  styleUrls: ['./autocomplete.component.scss'],
  providers: [AUTOCOMPLETE_VALUE_ACCESSOR],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
@AutoUnsubscribe()
export class AutocompleteComponent<ValueType> implements ControlValueAccessor {
  @Input()
  public value: ValueType | null = null;

  @Input()
  @CoerceBoolean()
  public required = false;

  @Input()
  public openOn: OpenOn = 'focus';

  @Input()
  @CoerceBoolean()
  public disabled?: boolean;

  @Output()
  public readonly valueChange: EventEmitter<ValueType | null> = new EventEmitter<ValueType | null>();

  @Input()
  public createOptionFn?: CreateOptionFn<ValueType>;

  @ViewChild(InputDirective, { static: true })
  private _input!: InputDirective;

  private _displayedValues$: BehaviorSubject<ValueType[]> = new BehaviorSubject<
    ValueType[]
  >(this.options);

  get displayedValues$(): Observable<ValueType[]> {
    return this._displayedValues$.asObservable();
  }

  private _options: ValueType[] = [];

  @Input()
  set options(options: ValueType[]) {
    this._options = (options || []).sort((a: ValueType, b: ValueType) =>
      this.displayOptionFn(a) < this.displayOptionFn(b) ? -1 : 1
    );

    this._displayedValues$.next(this._options);
  }

  get displayedValue(): string {
    return this.value !== null ? this.displayOptionFn(this.value) : '';
  }

  private _isOpen = false;

  get isOpen(): boolean {
    return this._isOpen;
  }

  constructor(
    private _elementRef: ElementRef,
    private _cdr: ChangeDetectorRef
  ) {}

  @Input()
  public displayOptionFn: DisplayFn<ValueType> = (option: ValueType): string =>
    option as unknown as string;

  @Input()
  public identityFn: IdentityFn<ValueType> = (value: ValueType): unknown =>
    value;

  public readonly trackByFn: TrackByFunction<ValueType> = (
    _: number,
    value: ValueType
  ) => this.identityFn(value);

  @HostListener('document:click', ['$event.target'])
  public onClick(event: EventTarget): void {
    if (!this._elementRef.nativeElement.contains(event)) {
      this._isOpen = false;
    }
  }

  public openPanel(): void {
    if (this.openOn === 'focus') {
      this._isOpen = true;
    }
  }

  public select(option: ValueType): void {
    if (
      this.value === null ||
      this.identityFn(this.value) !== this.identityFn(option)
    ) {
      this.propagateChange(option);
    } else if (!this.required) {
      this.propagateChange(null);
    }
  }

  public filterOptions(target: EventTarget | null): void {
    if (this.openOn === 'input' && !this.isOpen) {
      this._isOpen = true;
    }
    this._displayedValues$.next(
      this._options.filter((option: ValueType) =>
        this.displayOptionFn(option)
          .toLocaleLowerCase()
          .includes((target as HTMLInputElement).value.toLocaleLowerCase())
      )
    );
  }

  public isSelected(option: ValueType): boolean {
    return (
      this.value !== null &&
      this.identityFn(this.value) === this.identityFn(option)
    );
  }

  public writeValue(value: ValueType): void {
    this.value = value;
  }

  public setDisabledState(isDisabled: boolean): void {
    this.disabled = isDisabled;
    this._cdr.markForCheck();
  }

  public registerOnChange(fn: OnChangeFn<ValueType>): void {
    this._onChange = fn;
  }

  public registerOnTouched(fn: OnTouchedFn): void {
    this._onTouched = fn;
  }

  public createOption(target: EventTarget | null): void {
    const value: string = (target as HTMLInputElement).value;

    if (
      typeof this.createOptionFn === 'function' &&
      value.length !== 0 &&
      this._displayedValues$.getValue().length === 0
    ) {
      this.propagateChange(this.createOptionFn(value));
    }
  }

  protected _onChange: OnChangeFn<ValueType> = (): void => {
    // default if no ngControl
  };

  protected _onTouched: OnTouchedFn = (): void => {
    // default if no ngControl
  };

  private propagateChange(value: ValueType | null): void {
    this.value = value;
    this._isOpen = false;
    this._onChange(this.value);
    this._onTouched();
    this.valueChange.emit(this.value);
  }
}

@NgModule({
  declarations: [AutocompleteComponent],
  exports: [AutocompleteComponent],
  imports: [CommonModule, InputModule],
})
export class AutocompleteModule {}
