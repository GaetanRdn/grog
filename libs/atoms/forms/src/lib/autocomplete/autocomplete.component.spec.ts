import { Component, DebugElement } from '@angular/core';
import { TestBed } from '@angular/core/testing';
import { FormControl, ReactiveFormsModule } from '@angular/forms';
import { AutocompleteComponent, AutocompleteModule } from './autocomplete.component';
import { TemplateLookup } from '@grorg/tests';
import { CreateOptionFn, DisplayFn, IdentityFn } from './autocomplete.models';

describe('AutocompleteComponent', () => {
  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [
        HostComponent,
        RequiredHostComponent,
        ObjectValueHostComponent,
        OpenOnInputHostComponent,
        DisabledHostComponent,
        ReactiveFormHostComponent,
        AddToOptionsIfNotFoundHostComponent,
      ],
      imports: [AutocompleteModule, ReactiveFormsModule],
    });
  });

  describe('Basic', () => {
    let templateLookup: TemplateLookup<HostComponent>;

    beforeEach(() => {
      templateLookup = new TemplateLookup<HostComponent>(
        TestBed.createComponent(HostComponent)
      );

      templateLookup.detectChanges();
    });

    test('should create', () => {
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    test('when focus then options panel is open', () => {
      // WHEN
      templateLookup.query('input').focus();
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    test('when click on option then panel is closed', () => {
      // GIVEN
      templateLookup.query('input').focus();
      templateLookup.detectChanges();

      // WHEN
      templateLookup.query('.adr-option:last-child').click();
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    test('when click on option then check value', () => {
      // GIVEN
      templateLookup.query('input').focus();
      templateLookup.detectChanges();

      // WHEN
      templateLookup.query('.adr-option').click();
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.getComponent(AutocompleteComponent).value).toEqual(
        'Gaetan'
      );
      expect(templateLookup.hostComponent.value).toEqual('Gaetan');
    });

    test('when click on selected option then value is null', () => {
      // GIVEN
      templateLookup.query('input').focus();
      templateLookup.detectChanges();

      // WHEN
      templateLookup.query('.adr-option.adr-selected').click();
      templateLookup.detectChanges();

      // THEN
      expect(
        templateLookup.getComponent(AutocompleteComponent).value
      ).toBeNull();
      expect(templateLookup.hostComponent.value).toBeNull();
    });

    test('when click outside then options panel is closed', () => {
      // GIVEN
      templateLookup.query('input').focus();
      templateLookup.detectChanges();

      // WHEN
      templateLookup.query('.outside').click();
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    test('when input value then displayed options are filtered', () => {
      // GIVEN
      const input: DebugElement = templateLookup.get('input');
      input.nativeElement.value = 'g';
      input.triggerEventHandler('input', { target: input.nativeElement });

      // WHEN
      templateLookup.query('input').focus();
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });
  });

  describe('Required', () => {
    let templateLookup: TemplateLookup<RequiredHostComponent>;

    beforeEach(() => {
      templateLookup = new TemplateLookup<RequiredHostComponent>(
        TestBed.createComponent(RequiredHostComponent)
      );

      templateLookup.detectChanges();
    });

    test('should create', () => {
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    test('when required and click on selected option then value does not change', () => {
      // GIVEN
      jest.spyOn(
        templateLookup.getComponent(AutocompleteComponent).valueChange,
        'emit'
      );
      templateLookup.query('input').focus();
      templateLookup.detectChanges();

      // WHEN
      templateLookup.query('.adr-option.adr-selected').click();
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
      expect(
        templateLookup.getComponent(AutocompleteComponent).valueChange.emit
      ).not.toBeCalled();
      expect(templateLookup.getComponent(AutocompleteComponent).value).toEqual(
        'Soren'
      );
      expect(templateLookup.hostComponent.value).toEqual('Soren');
    });
  });

  describe('Objects as values', () => {
    let templateLookup: TemplateLookup<ObjectValueHostComponent>;

    beforeEach(() => {
      templateLookup = new TemplateLookup<ObjectValueHostComponent>(
        TestBed.createComponent(ObjectValueHostComponent)
      );

      templateLookup.detectChanges();
    });

    test('should create', () => {
      // WHEN
      templateLookup.query('input').focus();
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });
  });

  describe('Open on input', () => {
    let templateLookup: TemplateLookup<OpenOnInputHostComponent>;

    beforeEach(() => {
      templateLookup = new TemplateLookup<OpenOnInputHostComponent>(
        TestBed.createComponent(OpenOnInputHostComponent)
      );

      templateLookup.detectChanges();
    });

    test('should not open on focus', () => {
      // WHEN
      templateLookup.query('input').focus();
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    test('should open on input with value', () => {
      // GIVEN
      const input: DebugElement = templateLookup.get('input');
      input.nativeElement.value = 's';
      templateLookup.detectChanges();

      // WHEN
      input.triggerEventHandler('input', { target: input.nativeElement });
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });
  });

  describe('Disabled', () => {
    let templateLookup: TemplateLookup<DisabledHostComponent>;

    beforeEach(() => {
      templateLookup = new TemplateLookup<DisabledHostComponent>(
        TestBed.createComponent(DisabledHostComponent)
      );

      templateLookup.detectChanges();
    });

    test('should create', () => {
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    test('should not open on focus', () => {
      // WHEN
      templateLookup.query('input').focus();
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });
  });

  describe('Reactive form', () => {
    let templateLookup: TemplateLookup<ReactiveFormHostComponent>;

    beforeEach(() => {
      templateLookup = new TemplateLookup<ReactiveFormHostComponent>(
        TestBed.createComponent(ReactiveFormHostComponent)
      );

      templateLookup.detectChanges();
    });

    test('should create', () => {
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    test('should update control on select', () => {
      // GIVEN
      templateLookup.query('input').focus();
      templateLookup.detectChanges();
      expect(templateLookup.firstChildElement).toMatchSnapshot();

      // WHEN
      templateLookup.query('.adr-option:not(.adr-selected)').click();
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
      expect(templateLookup.hostComponent.control.value).toEqual({
        id: 2,
        firstName: 'Gaetan',
      });
    });

    test('should be disabled', () => {
      // WHEN
      templateLookup.hostComponent.control.disable();
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });
  });

  describe('Create option if not found', () => {
    let templateLookup: TemplateLookup<AddToOptionsIfNotFoundHostComponent>;

    beforeEach(() => {
      templateLookup = new TemplateLookup<AddToOptionsIfNotFoundHostComponent>(
        TestBed.createComponent(AddToOptionsIfNotFoundHostComponent)
      );

      templateLookup.detectChanges();
    });

    test('should add option', () => {
      // GIVEN
      const input: DebugElement = templateLookup.get('input');
      input.nativeElement.value = 'Jojo';

      // WHEN
      input.triggerEventHandler('focus', { target: input.nativeElement });
      templateLookup.detectChanges();
      input.triggerEventHandler('blur', { target: input.nativeElement });
      templateLookup.detectChanges();

      // THEN
      // expect(templateLookup.hostComponent.value).toEqual('Jojo');
    });
  });
});

interface Person {
  id: number;
  firstName: string;
}

@Component({
  template: ` <adr-autocomplete
      [options]="options"
      [(value)]="value"
    ></adr-autocomplete>
    <span class="outside"></span>`,
})
class HostComponent {
  public options: string[] = ['Soren', 'Gaetan'];

  public value: string = 'Soren';
}

@Component({
  template: ` <adr-autocomplete
    [options]="options"
    [(value)]="value"
    disabled
  ></adr-autocomplete>`,
})
class DisabledHostComponent {
  public options: string[] = ['Soren', 'Gaetan'];

  public value: string = 'Soren';
}

@Component({
  template: ` <adr-autocomplete
    [options]="options"
    [(value)]="value"
    required
  ></adr-autocomplete>`,
})
class RequiredHostComponent extends HostComponent {}

@Component({
  template: ` <adr-autocomplete
    [options]="options"
    [(value)]="value"
    openOn="input"
  ></adr-autocomplete>`,
})
class OpenOnInputHostComponent extends HostComponent {}

@Component({
  template: ` <adr-autocomplete
    [options]="options"
    [(value)]="value"
    [createOptionFn]="createOptionFn"
  ></adr-autocomplete>`,
})
class AddToOptionsIfNotFoundHostComponent extends HostComponent {
  public createOptionFn: CreateOptionFn<string> = (input: string) => input;
}

@Component({
  template: ` <adr-autocomplete
      [options]="options"
      [(value)]="value"
      [displayOptionFn]="displayFn"
      required
    ></adr-autocomplete>
    <span class="outside"></span>`,
})
class ObjectValueHostComponent {
  public options: Person[] = [
    { id: 1, firstName: 'Soren' },
    { id: 2, firstName: 'Gaetan' },
  ];

  public value: Person = { id: 1, firstName: 'Soren' };

  public displayFn: DisplayFn<Person> = (option: Person): string =>
    option.firstName;
}

@Component({
  template: ` <adr-autocomplete
      [options]="options"
      [formControl]="control"
      [displayOptionFn]="displayFn"
      [identityFn]="identityFn"
    ></adr-autocomplete>
    <span class="outside"></span>`,
})
class ReactiveFormHostComponent {
  public options: Person[] = [
    { id: 1, firstName: 'Soren' },
    { id: 2, firstName: 'Gaetan' },
  ];

  public readonly control: FormControl = new FormControl({
    id: 1,
    firstName: 'Soren',
  });

  public identityFn: IdentityFn<Person> = (p: Person) => p.id;

  public displayFn: DisplayFn<Person> = (option: Person): string =>
    option.firstName;
}
