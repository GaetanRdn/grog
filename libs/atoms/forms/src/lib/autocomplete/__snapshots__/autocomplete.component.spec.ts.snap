// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutocompleteComponent Basic should create 1`] = `
<adr-autocomplete>
  <input
    adrinput=""
    class="adr-input"
    value="Soren"
  />
  
</adr-autocomplete>
`;

exports[`AutocompleteComponent Basic when click on option then panel is closed 1`] = `
<adr-autocomplete>
  <input
    adrinput=""
    class="adr-input adr-focused"
    value=""
  />
  
</adr-autocomplete>
`;

exports[`AutocompleteComponent Basic when click outside then options panel is closed 1`] = `
<adr-autocomplete>
  <input
    adrinput=""
    class="adr-input adr-focused"
    value="Soren"
  />
  
</adr-autocomplete>
`;

exports[`AutocompleteComponent Basic when focus then options panel is open 1`] = `
<adr-autocomplete
  class="adr-opened"
>
  <input
    adrinput=""
    class="adr-input adr-focused"
    value="Soren"
  />
  
  <div
    class="adr-autocomplete-options"
  >
    
    <span
      class="adr-option"
    >
      Gaetan
    </span>
    <span
      class="adr-option adr-selected"
    >
      Soren
    </span>
  </div>
</adr-autocomplete>
`;

exports[`AutocompleteComponent Basic when input value then displayed options are filtered 1`] = `
<adr-autocomplete
  class="adr-opened"
>
  <input
    adrinput=""
    class="adr-input adr-focused"
    value="g"
  />
  
  <div
    class="adr-autocomplete-options"
  >
    
    <span
      class="adr-option"
    >
      Gaetan
    </span>
  </div>
</adr-autocomplete>
`;

exports[`AutocompleteComponent Disabled should create 1`] = `
<adr-autocomplete
  class="adr-disabled"
  disabled="true"
>
  <input
    adrinput=""
    class="adr-input adr-disabled"
    disabled=""
    value="Soren"
  />
  
</adr-autocomplete>
`;

exports[`AutocompleteComponent Disabled should not open on focus 1`] = `
<adr-autocomplete
  class="adr-disabled"
  disabled="true"
>
  <input
    adrinput=""
    class="adr-input adr-disabled"
    disabled=""
    value="Soren"
  />
  
</adr-autocomplete>
`;

exports[`AutocompleteComponent Objects as values should create 1`] = `
<adr-autocomplete
  class="adr-opened"
  required="true"
>
  <input
    adrinput=""
    class="adr-input adr-focused"
    value="Soren"
  />
  
  <div
    class="adr-autocomplete-options"
  >
    
    <span
      class="adr-option"
    >
      Soren
    </span>
    <span
      class="adr-option"
    >
      Gaetan
    </span>
  </div>
</adr-autocomplete>
`;

exports[`AutocompleteComponent Open on input should not open on focus 1`] = `
<adr-autocomplete
  openon="input"
>
  <input
    adrinput=""
    class="adr-input adr-focused"
    value="Soren"
  />
  
</adr-autocomplete>
`;

exports[`AutocompleteComponent Open on input should open on input with value 1`] = `
<adr-autocomplete
  class="adr-opened"
  openon="input"
>
  <input
    adrinput=""
    class="adr-input"
    value="s"
  />
  
  <div
    class="adr-autocomplete-options"
  >
    
    <span
      class="adr-option adr-selected"
    >
      Soren
    </span>
  </div>
</adr-autocomplete>
`;

exports[`AutocompleteComponent Reactive form should be disabled 1`] = `
<adr-autocomplete
  class="ng-untouched ng-pristine adr-disabled"
  disabled="true"
>
  <input
    adrinput=""
    class="adr-input adr-disabled"
    disabled=""
    value="Soren"
  />
  
</adr-autocomplete>
`;

exports[`AutocompleteComponent Reactive form should create 1`] = `
<adr-autocomplete
  class="ng-untouched ng-pristine ng-valid"
>
  <input
    adrinput=""
    class="adr-input"
    value="Soren"
  />
  
</adr-autocomplete>
`;

exports[`AutocompleteComponent Reactive form should update control on select 1`] = `
<adr-autocomplete
  class="ng-untouched ng-pristine ng-valid adr-opened"
>
  <input
    adrinput=""
    class="adr-input adr-focused"
    value="Soren"
  />
  
  <div
    class="adr-autocomplete-options"
  >
    
    <span
      class="adr-option adr-selected"
    >
      Soren
    </span>
    <span
      class="adr-option"
    >
      Gaetan
    </span>
  </div>
</adr-autocomplete>
`;

exports[`AutocompleteComponent Reactive form should update control on select 2`] = `
<adr-autocomplete
  class="ng-valid ng-touched ng-dirty"
>
  <input
    adrinput=""
    class="adr-input adr-focused"
    value="Gaetan"
  />
  
</adr-autocomplete>
`;

exports[`AutocompleteComponent Required should create 1`] = `
<adr-autocomplete
  required="true"
>
  <input
    adrinput=""
    class="adr-input"
    value="Soren"
  />
  
</adr-autocomplete>
`;

exports[`AutocompleteComponent Required when required and click on selected option then value does not change 1`] = `
<adr-autocomplete
  class="adr-opened"
  required="true"
>
  <input
    adrinput=""
    class="adr-input adr-focused"
    value="Soren"
  />
  
  <div
    class="adr-autocomplete-options"
  >
    
    <span
      class="adr-option"
    >
      Gaetan
    </span>
    <span
      class="adr-option adr-selected"
    >
      Soren
    </span>
  </div>
</adr-autocomplete>
`;
